### PR TITLE
Added an explanation of how to use events to the create-window example

### PR DESCRIPTION
**Pull request by scrouthtv**
Hello,

first, I think that the default examples are very helpful and well documented.

However, if you run the create-window example and close the window, the underlying Go Program keeps running.

I think this is a good use case for the `DestroyNotifyEvent` and demonstrating how reading it / reading events in general works.

So I added these simple lines to the demo.

### CODE REVIEW

File: examples/create-window/main.go [hunk 1]
```diff
@@ -96,12 +96,22 @@ func main() {
...
 			fmt.Printf("Error: %s\n", xerr)
 		}
+		
+		// This is how accepting events work:
+		// The application checks what event we got
+		// (the event must be registered using either xproto.CreateWindow (see l.35) or
+		//  xproto.ChangeWindowAttributes (see l.50)).
+		// and reacts to it accordingly. All events are defined in the xproto subpackage.
+		switch ev.(type) {
+		case xproto.DestroyNotifyEvent:
```
Comment (jezek) [inline examples/create-window/main.go:113]: Someone might wonder, why the DestroyNotifyEvent won't trigger if the window is closed by user (at least on Ubuntu). Does it trigger on window close for you?

I see, you mentioned it in the comment above (must be registered...), but is it ok to have unreachable code in example? I mean If we show B (how to catch events), don't we have to show A (how to register)?

File: examples/create-window/main.go [hunk 2]
```diff
@@ -96,12 +96,22 @@ func main() {
 			fmt.Println("Both event and error are nil. Exiting...")
 			return
 		}
-
+
```
Comment (jezek) [inline examples/create-window/main.go:99]: Please, always format go code with ```go fmt```. If you are using an IDE, configure it to run formatting after save.

File: examples/create-window/main.go [hunk 3]
```diff
@@ -96,12 +96,22 @@ func main() {
...
 		}
-
+		
 		if ev != nil {
 			fmt.Printf("Event: %s\n", ev)
 		}
 		if xerr != nil {
 			fmt.Printf("Error: %s\n", xerr)
 		}
+
```
Comment (jezek) [inline examples/create-window/main.go:106]: Another formatting mistake, which will be fixed by ```go fmt```.

**Review by jezek — COMMENTED** [review]
> However, if you run the create-window example and close the window, the underlying Go Program keeps running.

How running? On Ubuntu if I close the window, i got the program terminated with output:
```
XGB: xgb.go:455: A read error is unrecoverable: EOF
Both event and error are nil. Exiting...
```

**Comment by scrouthtv** [comment]
I just tried again:
xgb-pr1
(It only ends because I manually C-c)

I am running Arch Linux with `xorg-server 1.20.10-3` and `xcb-proto 1.14.1-3`. My WM is `i3-gaps 4.19.1-1`.

If I close the window, both UnmapNotify and DestroyNotify are sent, however, no error is sent (which closes the example in your case).

Sorry about the formatting.

**Comment by scrouthtv** [comment]
Here is also an example of how to listen to key

**Comment by jezek** [comment]
Now I remember. I've encountered this before. On some (well behaved) DEs the UnmapNotify and DestroyNotify is not sent just the connection is closed, unless you [add WM_DELETE_WINDOW to WM_PROTOCOLS property](https://github.com/jezek/xgbutil/blob/master/_examples/graceful-window-close/main.go). Could you mention this in comment in the DestoyNotifyEvent, please?

**Comment by scrouthtv** [comment]
sure thing. You're using stock ubuntu?

I also have an example for drawing coming up which you can consider adding, I just have to add comments for everything.

File: examples/create-window/main.go [hunk 4]
```diff
@@ -17,6 +17,7 @@ func main() {
 		fmt.Println(err)
 		return
 	}
+	defer X.Close()
```
Comment (scrouthtv) [inline examples/create-window/main.go:20]: Also put an `X.Close()` in there for good measure. What do you think?

**Comment by scrouthtv** [comment]
Here it is. I think I'm done

**Comment by scrouthtv** [comment]
I also added some text drawing in there.

File: examples/create-window/main.go (hunk 4 above)
Comment (jezek) [reply examples/create-window/main.go:20]: Yeah, this is a good practice.

File: examples/create-window/main.go [hunk 5]
```diff
@@ -103,5 +105,40 @@ func main() {
 		if xerr != nil {
 			fmt.Printf("Error: %s\n", xerr)
 		}
+
+		// This is how accepting events work:
+		// The application checks what event we got
+		// (the event must be registered using either xproto.CreateWindow (see l.35) or
+		//  xproto.ChangeWindowAttributes (see l.50)).
```
Comment (jezek) [inline examples/create-window/main.go:112]: Please reword like this
```
The application checks what event we got and reacts to it accordingly. All events are defined in the xproto subpackage and the event we wan't to handle must be registered using either xproto.CreateWindow or xproto.ChangeWindowAttributes before.
```

Why? I don't think it's good practice to mention lines in comments. After changes you'll have to check comments if lines are matching and make fixes. Also too long text in parenthesis and nested () are a distraction.

File: examples/create-window/main.go [hunk 6]
```diff
@@ -103,5 +105,40 @@ func main() {
...
+		// and reacts to it accordingly. All events are defined in the xproto subpackage.
+		switch ev.(type) {
+		case xproto.KeyPressEvent:
+			// See https://pkg.go.dev/github.com/jezek/xgb@v0.0.0-20210121230032-cec22bda1ce1/xproto#KeyPressEvent
```
Comment (jezek) [inline examples/create-window/main.go:116]: Please, use ```https://pkg.go.dev/github.com/jezek/xgb/xproto#KeyPressEvent``` for better readability.

File: examples/create-window/main.go [hunk 7]
```diff
@@ -103,5 +105,40 @@ func main() {
...
+				return // exit on q
+			}
+		case xproto.DestroyNotifyEvent:
+			// Depending on the user's desktop environment (especially
+			// window manager), killing a window might close the
+			// client's X connection (e. g. the default Ubuntu
+			// desktop environment).
+			//
+			// If that's the case for your environment, closing this example's window
+			// will also call the underlying Go program (because closing the X
```
Comment (jezek) [inline examples/create-window/main.go:132]: Shouldn't the ```call``` be ```kill```?

File: examples/create-window/main.go [hunk 8]
```diff
@@ -103,5 +105,40 @@ func main() {
...
+			// connection gives a nil event and EOF error).
```
Comment (jezek) [inline examples/create-window/main.go:133]: Could you mention, that an example with graceful close is [here](https://github.com/jezek/xgbutil/blob/master/_examples/graceful-window-close/main.go), please?

File: examples/shapes/main.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
+// Example shapes shows how to draw basic shapes into a window.
```
Comment (jezek) [inline examples/shapes/main.go:1]: Could you move this example to new PR please? This one is called ```Added an explanation of how to use events to the create-window example``` and should contain only changes related to this PR.

**Review by jezek — CHANGES_REQUESTED** [review]
It looks a lot better, but I have a few demands. Thank you for bearing with me and implementing my requests so far.

> You're using stock ubuntu (so I can mention your DE as an example for closing = killing the X conection)?

I use Ubuntu MATE right now, but I think, stock ubuntu does the same. ~~I have to test it, but for now let's assume it's happening on stock ubuntu too. :laughing:~~
Edit: Tested it on stock Ubuntu 20.04, no DestroyNotifyEvent, like in MATE.

File: examples/shapes/main.go (hunk 9 above)
Comment (scrouthtv) [reply examples/shapes/main.go:1]: #3

File: examples/create-window/main.go (hunk 8 above)
Comment (scrouthtv) [reply examples/create-window/main.go:133]: Sure thing. 28ff8ce

File: examples/create-window/main.go (hunk 7 above)
Comment (scrouthtv) [reply examples/create-window/main.go:132]: Oh yeah, typo on my side. Sorry for that

File: examples/create-window/main.go (hunk 8 above)
Comment (scrouthtv) [reply examples/create-window/main.go:133]: 257276c

File: examples/shapes/main.go (hunk 9 above)
Comment (scrouthtv) [reply examples/shapes/main.go]: 257276c

File: examples/create-window/main.go (hunk 6 above)
Comment (scrouthtv) [reply examples/create-window/main.go]: 257276c

File: examples/create-window/main.go (hunk 5 above)
Comment (scrouthtv) [reply examples/create-window/main.go]: c3652e2

**Comment by scrouthtv** [comment]
> Thank you for bearing with me

No worries, I am a perfectionist myself ;)

I think it should all be in there.
Regarding the shapes example, I messed up a bit and will be opening a new PR soon (in case you're interested).

**Comment by jezek** [comment]
Just added an missing "\n" and merged. Thank you.

Also I'm interested in the shapes example. I myself like to learn from examples, so I think the more examples, the beter. ;)
Hint: ```git rebase -i``` is useful to trim/modify branches.

**Pull request merged**